shims/common/src/main/scala/org/apache/gluten/config/ConfigEntry.scala

```diff
@@ -0,0 +1,203 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.config
+
+import org.apache.gluten.config.BackendType.BackendType
+
+trait ConfigEntry[T] {
+  def key: String
+  def doc: String
+  def version: String
+  def backend: BackendType.BackendType
+  def isPublic: Boolean
+  def alternatives: List[String]
+  def valueConverter: String => T
+  def stringConverter: T => String
+  def readFrom(conf: ConfigProvider): T
+  def defaultValue: Option[T]
+  def defaultValueString: String
+
+  final protected def readString(provider: ConfigProvider): Option[String] = {
+    alternatives.foldLeft(provider.get(key))((res, nextKey) => res.orElse(provider.get(nextKey)))
+  }
+
+  override def toString: String = {
+    s"ConfigEntry(key=$key, defaultValue=$defaultValueString, doc=$doc, " +
+      s"public=$isPublic, version=$version)"
+  }
+
+  ConfigEntry.registerEntry(this)
+}
+
+private[gluten] class OptionalConfigEntry[T](
+    _key: String,
+    _doc: String,
+    _version: String,
+    _backend: BackendType,
+    _isPublic: Boolean,
+    _alternatives: List[String],
+    _valueConverter: String => T,
+    _stringConverter: T => String)
+  extends ConfigEntry[Option[T]] {
+  override def key: String = _key
+
+  override def doc: String = _doc
+
+  override def version: String = _version
+
+  override def backend: BackendType = _backend
+
+  override def isPublic: Boolean = _isPublic
+
+  override def alternatives: List[String] = _alternatives
+
+  override def valueConverter: String => Option[T] = s => Option(_valueConverter(s))
+
+  override def stringConverter: Option[T] => String = v => v.map(_stringConverter).orNull
+
+  override def readFrom(conf: ConfigProvider): Option[T] = readString(conf).map(_valueConverter)
+
+  override def defaultValue: Option[Option[T]] = None
+
+  override def defaultValueString: String = ConfigEntry.UNDEFINED
+}
+
+private[gluten] class ConfigEntryWithDefault[T](
+    _key: String,
+    _doc: String,
+    _version: String,
+    _backend: BackendType,
+    _isPublic: Boolean,
+    _alternatives: List[String],
+    _valueConverter: String => T,
+    _stringConverter: T => String,
+    _defaultVal: T)
+  extends ConfigEntry[T] {
+  override def key: String = _key
+
+  override def doc: String = _doc
+
+  override def version: String = _version
+
+  override def backend: BackendType = _backend
+
+  override def isPublic: Boolean = _isPublic
+
+  override def alternatives: List[String] = _alternatives
+
+  override def valueConverter: String => T = _valueConverter
+
+  override def stringConverter: T => String = _stringConverter
+
+  override def readFrom(conf: ConfigProvider): T = {
+    readString(conf).map(valueConverter).getOrElse(_defaultVal)
+  }
+
+  override def defaultValue: Option[T] = Option(_defaultVal)
+
+  override def defaultValueString: String = stringConverter(_defaultVal)
+}
+
+private class ConfigEntryWithDefaultString[T](
```

Comment:
This one is `private` (actually equivalent to `private[config]`) while others are `private[gluten]`. Is there any particular reason?

Reply:
no, will keep consistent.